Overhaul of MediaDownloader: simpler, handles `Content-Encoding: gzip`

mmdriley opened the pull request:
Today, MediaDownloader downloads content in chunks by sending one HTTP
request per chunk. Each request has an appropriate `Range` header set.

One consequence of this approach is that when users try to download
content served with `Content-Encoding: gzip` and that content happens
to be more than one chunk large, the download will fail with an error
like:

`System.IO.InvalidDataException: The magic number in GZip header is not correct. Make sure you are passing in a GZip stream.`

`HttpClientHandler` is trying to do the decoding for us because we set
its `AutomaticDecompression` property, but the decoding fails because
we've requested a range of data in the middle of a GZip stream.

Users can encounter this when downloading files from GCS that were
uploaded with `gsutil -Z`.

I considered asking the server to send us already-decompressed content
by not setting the `Accept-Encoding` header in our requests. Aside from
wasting  customers' bandwidth, this turns out not to work: GCS ignores
the header and returns gzipped data regardless.

I also briefly considered special case handling for GZip downloads,
e.g. setting a huge ChunkSize or buffering intermediate results.

Instead, I decided to make things simpler and more flexible by only
making one request per download. As far as I can tell, there wasn't any
benefit to the multiple-request approach. It probably hurt performance
because we interrupted established transfers.

To avoid any impact on clients, ChunkSize is retained. Now it indicates
the granularity at which content should be written to the output stream
and our callers will be notified of progress. I've verified that the
progress events the caller sees are the same for both implementations.

I could not find a way to test this in MediaDownloaderTest as it was.
We were mocking out the request side of HttpClient and not exercising
any of the code that is actually responsible for transport concerns
like compression. My solution was to rewrite MediaDownloaderTest to
use a local HTTP server that serves the responses we need to exercise
the behavior under test. A lot of code changed, but I really think the
resulting test is easier to read and less fragile than before.

The old MediaDownloader implementation passes the new tests (except
the one that returns GZip content). Only minimal modifications to the
old tests are necessary for them to pass on the new MediaDownloader --
mostly removing those tests' baked-in assumptions about how many
requests would be made and for what ranges.

While I was there, I replaced some fiddly Query string manipulation
code in MediaDownloader that took apart the Query string only to
immediately reassembly it. That removed MediaDownloader's use of
RequestBuilder. Since the latter is what we trusted to bring in our
URI hacks, I made `PatchUriQuirks` public and added a class
initializer to MediaDownloader to use it.

mmdriley commented:
@jskeet @peleyal 

Apologies for the large PR. I found it's much easier to read as a "Split" diff.

I'm out for the rest of the week; I'll work on addressing feedback when I'm back on Monday.

mmdriley commented:
FYI, when I say "minimal modifications to the old tests", here's what I had to change to get them to pass with the new implementation: https://github.com/google/google-api-dotnet-client/compare/master...mmdriley:old-test-new-code#diff-a4fca60bfee5ead670867ffff7e7d3ddL90

jskeet commented on Src/Support/GoogleApis/Apis/[Media]/Download/MediaDownloader.cs:
```diff
@@ -166,96 +175,86 @@ public IDownloadProgress Download(string url, Stream stream)
-            if (string.IsNullOrEmpty(uri.Query))
+            // Add alt=media to the query parameters.
+            var uri = new UriBuilder(url);
+            if (uri.Query == null || uri.Query.Length <= 1)
```
The change to this condition suggests that we sometimes see a single-character query here, which we want to replace. What's the background to this?

mmdriley replied:
I'm deferring to [the docs for UriBuilder.Query](https://msdn.microsoft.com/en-us/library/system.uribuilder.query(v=vs.110).aspx), which say:
> Do not append a string directly to this property. If the length of Query is greater than 1, retrieve the property value as a string, remove the leading question mark, append the new query string, and set the property with the combined string.

I don't believe I actually saw just "?" in my testing.

jskeet commented on Src/Support/GoogleApis/Apis/[Media]/Download/MediaDownloader.cs:
```diff
@@ -148,8 +153,12 @@ public IDownloadProgress Download(string url, Stream stream)
+        /// 
+        /// The chunking behavior is largely a historical artifact: a previous implementation
+        /// issued multiple web requests, each for ChunkSize bytes. Now we do everything in
+        /// one request, but the API and client-visible behavior are retained for compatibility.
```
Will we still actually get large files in chunks, so that meaningful progress is reported, and proper streaming behaviour is observed? (I wouldn't want this change to mean that something inside .NET buffered the whole file in memory, for example.)

mmdriley replied:
Yes, as before we buffer only up to ChunkSize bytes in memory before writing them to the client's stream.

jskeet commented on Src/Support/GoogleApis/Apis/[Media]/Download/MediaDownloader.cs:
```diff
@@ -166,96 +175,86 @@ public IDownloadProgress Download(string url, Stream stream)
+                                bufferOffset += read;
+                            }
+
+                            await stream.WriteAsync(buffer, 0, bufferOffset, cancellationToken).ConfigureAwait(false);
```
Is this valid if bufferOffset is 0?

mmdriley replied:
I think so. I don't see anything [in the docs](https://msdn.microsoft.com/en-us/library/hh158629(v=vs.110).aspx) that suggests otherwise.

jskeet commented on Src/Support/GoogleApis/Apis/[Media]/Download/MediaDownloader.cs:
```diff
@@ -166,96 +175,86 @@ public IDownloadProgress Download(string url, Stream stream)
+                            // If we read less than ChunkSize, we've reached the end of the stream.
+                            // In that case, avoid reporting a Downloading event since we'll instead
+                            // report Completed.
+                            if (bufferOffset < buffer.Length)
```
That suggests that a file with a length which is exactly a multiple of ChunkSize will report a final Downloading with the final size, whereas anything else won't. I think I'm okay with that, but I just wanted to check.

mmdriley replied:
Hm, sure enough. That doesn't match the previous behavior. I'll look for a way to avoid this without mangling the code too much.

mmdriley replied:
Now fixed, with new code in the test to verify.

jskeet commented on Src/Support/GoogleApis.Tests/Apis/Download/MediaDownloaderTest.cs:
```diff
@@ -39,95 +39,154 @@ namespace Google.Apis.Tests.Apis.Download
+            try
+            {
+                tcpListener.Start();
+                return (tcpListener.LocalEndpoint as IPEndPoint).Port;
```
Tiny nit: I'd prefer a cast here, so that if it's ever *not* an IPEndPoint, we'll get a InvalidCastException instead of NullReferenceException. Just one of my bugbears :)

mmdriley replied:
Done.

jskeet commented on Src/Support/GoogleApis.Tests/Apis/Download/MediaDownloaderTest.cs:
```diff
@@ -39,95 +39,154 @@ namespace Google.Apis.Tests.Apis.Download
+            _httpPrefix = String.Format(PrefixFormat, GetOpenPort());
+            Task.Run(() => RunTestServer(_httpPrefix));
+
+            // TODO: The server is only created once, but it still might be polite to tear it down
```
Easy to do in the "everything is working" case by adding a "Quit" URL, right?

mmdriley replied:
Added.

jskeet commented on Src/Support/GoogleApis.Tests/Apis/Download/MediaDownloaderTest.cs:
```diff
@@ -197,36 +266,51 @@ public void DownloadAsync_Cancel()
-        private void Subtest_Download_Chunks(int chunkSize, bool sync = true, int cancelRequest = 0,
-            string downloadUri = "http://www.sample.com")
+        /// <param name="target">Last component of the Uri to download</param>
+        private void Subtest_Download_Chunks(int chunkSize, bool sync = true, int cancelRequest = 0, string target = "content")
```
Possibly rename cancelRequest to cancelChunk as there's really only one request now?

mmdriley replied:
Done.

jskeet commented:
Love the new testing approach. Much nicer.

jskeet commented:
(Oh, and LGTM in general.)

mmdriley commented:
Many thanks! PTAL.

Adding the `/Quit` handler made me want async `TearDownTestFixture`, which led me to updating our tests to use NUnit 3. I expect to leave that as a separate commit in this PR.

Other changes are in the "PR feedback" commit, including:
- Added the /Quit handler.
- Fixed the redundant Downloading and Completed notification when the response size is divisible by ChunkSize. The `CountedBuffer` utility class is admittedly a bit verbose, though overall I found it made things more readable. Let me know if you don't find that the case.
- Added an assertion that BytesDownloaded always increases. The old implementation passed this test; the new implementation does now.

I plan to squash the "PR feedback" commit before merging.

jskeet commented:
LGTM. I tried to think of a simpler way of handling it, but couldn't come up with anything. The data copying (in RemoveFromFront) bothered me a little until I realized it would usually only be copying a single byte anyway...
